### PR TITLE
Add codebashing lessons per CWE

### PR DESCRIPTION
**Pull request by AvivCx**
### Description

Add CodeBashing integration to support direct link for tenant account to specific relevant lesson
Now the 'Training' link in cxflow will direct the user to specific lesson (if exist) - relevant to the vulnerability type

### References

https://github.com/checkmarx-ltd/cx-flow/wiki

### Checklist

- [x] I have added documentation for new/changed functionality in this PR (if applicable).  *If documentation is a Wiki Update, please indicate desired changes within PR MD Comment*
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used

### CODE REVIEW

File: src/main/java/com/checkmarx/flow/service/CodeBashingService.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            ResponseEntity<String> response = restTemplate.exchange(codebashingProperties.getCodebashingApiUrl(), HttpMethod.GET, httpEntity, String.class);
+
+            if (response.getBody()==null){
+             log.error("can't get codebashing lessons. response is null");
+            }
+
+            log.info("codebashing API get lessons response: {} - {}", response.getStatusCode(), response.getStatusCodeValue());
+
+            JSONArray lessonsArray = new JSONArray(response.getBody());
+            lessonsMap = createLessonMapByCwe(lessonsArray);
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/CodeBashingService.java:45]: Here a state is introduced inside a service, which is generally not a good thing. This map should be returned to the outside and not stored inside this service instance.

File: src/main/java/com/checkmarx/flow/service/IssueService.java [hunk 2]
```diff
@@ -112,6 +116,8 @@ public IssueService(FlowProperties properties) {
                     String fileUrl;
                     ScanResults.XIssue currentIssue = xIssue.getValue();
 
+                    codeBashingService.addCodebashingUrlToIssue(currentIssue);
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/IssueService.java:119]: Here we rely on codeBashingService internal state, as created by createLessonsMap. This should be fixed - please see my note above.

File: src/main/java/com/checkmarx/flow/service/JiraService.java [hunk 3]
```diff
@@ -74,12 +75,13 @@
     private static final int MAX_RESULTS_ALLOWED = 1000000;
     private static final String SEARCH_ASSIGNABLE_USER = "%s/rest/api/latest/user/assignable/search?project={projectKey}&query={assignee}";
 
-    @ConstructorProperties({"jiraProperties", "flowProperties"})
-    public JiraService(JiraProperties jiraProperties, FlowProperties flowProperties) {
+    @ConstructorProperties({"jiraProperties", "flowProperties", "codeBashingService"})
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/JiraService.java:78]: I believe this line can be safely removed.

File: src/main/java/com/checkmarx/flow/service/JiraService.java [hunk 4]
```diff
@@ -945,7 +947,7 @@ private String getBody(ScanResults.XIssue issue, ScanRequest request, String fil
         StringBuilder body = new StringBuilder();
         Optional.ofNullable(jiraProperties.getDescriptionPrefix())
                 .ifPresent(body::append);
-
+        Map<String, Object> additionalDetails = issue.getAdditionalDetails();
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/JiraService.java:950]: This declaration should be moved closer to its usage.

File: src/main/java/com/checkmarx/flow/service/JiraService.java [hunk 5]
```diff
@@ -1177,10 +1182,12 @@ private void isFileUrlEmpty(ScanRequest request, String fileUrl, StringBuilder b
         setMapWithScanResults(map, nonPublishedScanResultsMap);
         jiraMap = this.getJiraIssueMap(this.getIssues(request));
 
+
         for (Map.Entry<String, ScanResults.XIssue> xIssue : map.entrySet()) {
             String issueCurrentKey = xIssue.getKey();
             try {
                 ScanResults.XIssue currentIssue = xIssue.getValue();
+                codeBashingService.addCodebashingUrlToIssue(currentIssue);
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/JiraService.java:1190]: Relying on internal service state - similar problem as in IssueService.

File: src/main/java/com/checkmarx/flow/utils/HTMLHelper.java [hunk 6]
```diff
@@ -369,6 +370,8 @@ private static void setSCAHtmlBody(ScanResults.XIssue issue, ScanRequest request
     private static void setSASTMDBody(ScanResults.XIssue issue, String branch, String fileUrl,
             FlowProperties flowProperties, StringBuilder body) {
         log.debug("Building MD body for SAST scanner");
+        Map<String, Object> additionalDetails = issue.getAdditionalDetails();
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/utils/HTMLHelper.java:373]: Please move the declaration closer to its usage.

File: src/test/java/com/checkmarx/flow/cucumber/integration/azure/publishing/githubflow/ScanResultsBuilder.java [hunk 7]
```diff
@@ -50,7 +50,7 @@ private static CxScanSummary getScanSummary() {
                 .file("DOS_Login.java")
                 .description("Description: Reflected_XSS_All_Clients")
                 .details(getIssueDetails())
-                .additionalDetails(Collections.emptyMap())
+                .additionalDetails(map)
```
Comment (alex-ko-dev) [inline src/test/java/com/checkmarx/flow/cucumber/integration/azure/publishing/githubflow/ScanResultsBuilder.java:53]: `new HashMap<>()` can be inlined here.

File: src/test/java/com/checkmarx/flow/cucumber/integration/azure/publishing/githubflow/ScanResultsBuilder.java [hunk 8]
```diff
@@ -73,7 +73,7 @@ private static CxScanSummary getScanSummary() {
                 .file("DOS_Login.java")
                 .description("Description: SQL_Injection")
                 .details(getIssueDetails())
-                .additionalDetails(Collections.emptyMap())
+                .additionalDetails(map)
```
Comment (alex-ko-dev) [inline src/test/java/com/checkmarx/flow/cucumber/integration/azure/publishing/githubflow/ScanResultsBuilder.java:76]: `new HashMap<>()` can be inlined here.

**Review by alex-ko-dev — CHANGES_REQUESTED** [review]
Currently the major issue is storing state inside CodeBashingService - please see the comments.
Didn't have time to review the tests yet...

File: docs/CodeBashing.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
+## Integration with CodeBashing
+
+CxFlow adds CodeBashing link when creating a Bug-Tracker ticket.
+The link to CodeBasing is added under '**Training**' in the ticket recommendation:
```
[the remark this answers is not in the document]
Comment (AvivCx) [reply docs/CodeBashing.md]: Fixed

File: docs/CodeBashing.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+cx-flow:
+  codebash-url: https:...
+```
+
+Now each customer (CodeBashing tenant), can use cxflow to add training link to his tickets - with direct link to relevant course per language and CWE!
+The link will direct to the specific lesson under the customer CodeBashing account.
+for example:  (https://cxflow-account.codebashing.com/courses/backend_java/lessons/sql_injection)
+
+
+To get direct CodeBashing integration and connect ot to  your own tenant, add the following section to cxflow configuration:
```
[the remark this answers is not in the document]
Comment (AvivCx) [reply docs/CodeBashing.md]: Fixed

File: src/main/java/com/checkmarx/flow/config/CodebashingProperties.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.stereotype.Component;
+import org.springframework.validation.annotation.Validated;
+
+@Component
+@ConfigurationProperties(prefix = "codebashing")
+@Validated
+public class CodebashingProperties {
```
[the remark this answers is not in the document]
Comment (AvivCx) [reply src/main/java/com/checkmarx/flow/config/CodebashingProperties.java:16]: added to the class and removed from members

File: src/main/java/com/checkmarx/flow/service/CodeBashingService.java [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+import org.springframework.web.client.RestTemplate;
+import javax.validation.ValidationException;
+import java.util.HashMap;
+import java.util.Map;
+
+@RequiredArgsConstructor
+@Service
+public class CodeBashingService {
+
+    private static Logger log = org.slf4j.LoggerFactory.getLogger(CodeBashingService.class);
```
[the remark this answers is not in the document]
Comment (AvivCx) [reply src/main/java/com/checkmarx/flow/service/CodeBashingService.java:24]: added

File: src/main/java/com/checkmarx/flow/service/CodeBashingService.java [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    private Map<String,String> lessonsMap = null;
```
[the remark this answers is not in the document]
Comment (AvivCx) [reply src/main/java/com/checkmarx/flow/service/CodeBashingService.java]: removed

File: src/main/java/com/checkmarx/flow/service/JiraService.java (hunk 4 above)
Comment (AvivCx) [reply src/main/java/com/checkmarx/flow/service/JiraService.java]: Fixed

File: src/main/java/com/checkmarx/flow/utils/HTMLHelper.java (hunk 6 above)
Comment (AvivCx) [reply src/main/java/com/checkmarx/flow/utils/HTMLHelper.java]: moved

File: src/test/java/com/checkmarx/flow/cucumber/integration/azure/publishing/githubflow/ScanResultsBuilder.java (hunk 7 above)
Comment (AvivCx) [reply src/test/java/com/checkmarx/flow/cucumber/integration/azure/publishing/githubflow/ScanResultsBuilder.java]: Fixed

File: src/test/java/com/checkmarx/flow/cucumber/integration/azure/publishing/githubflow/ScanResultsBuilder.java (hunk 8 above)
Comment (AvivCx) [reply src/test/java/com/checkmarx/flow/cucumber/integration/azure/publishing/githubflow/ScanResultsBuilder.java]: Fixed

File: src/main/java/com/checkmarx/flow/service/CodeBashingService.java [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+        httpHeaders.set("x-api-key", codebashingProperties.getApiSecret());
+        return httpHeaders;
+    }
+
+    private void validateCodeBashingLessonsIntegration() {
+         if(codebashingProperties == null ||
+                 codebashingProperties.getCodebashingApiUrl() == null ||
+                 codebashingProperties.getTenantBaseUrl() == null ||
+                 codebashingProperties.getApiSecret() == null) {
+             throw new ValidationException();
```
[the remark this answers is not in the document]
Comment (AvivCx) [reply src/main/java/com/checkmarx/flow/service/CodeBashingService.java:123]: Added

**Pull request merged**